### PR TITLE
Solving the issue "`sort-by` should fail gracefully if mismatched types are compared"

### PR DESCRIPTION
**Pull request by luccasmmg**
This Pull Request intends to solve the issue https://github.com/nushell/nushell/issues/2357 This is my first pull request in a Open Source project controlled by someone else, so i would appreciate advice.
The issue talked about returning an error instead of using expect() the problem that i encountered was that the closure inside the sort_by has to return a enum:cmp i tried to somehow catch the panic but that didn't work also, so what i decided to do was make the coerce check before the sort, i could probably make this already change the values to be coerced, but i didn't want to mess up anything, so i made the new lines just check and return an error, which can probably be considered more convoluted.
I would like to know if anyone thought of a better solution, or if i should just tried to convert all the elems instead of just checking.
ps: sorry for my english

### CODE REVIEW

File: crates/nu-cli/src/commands/sort_by.rs [hunk 1]
```diff
@@ -135,6 +135,15 @@ pub fn sort(
         } => {
             let should_sort_case_insensitively = insensitive && vec.iter().all(|x| x.is_string());
 
+            if !vec
+                .windows(2)
+                .all(|elem| coerce_compare(&elem[0], &elem[1]).is_ok())
+            {
+                return Err(ShellError::labeled_error("Can't reduce all the values to all the values to the same type",
+                                                        "All these values should have some relation between then such that the shell knows how to compare them",
+                                                        tag));
```
Comment (thegedge) [inline crates/nu-cli/src/commands/sort_by.rs:144]: A few things:

1. style nit: the indentation is a little hard to follow here, and also puts the values far to the right.
2. You've repeated "all the values" in the messae. I think you want something like "Can't reduce all of the values to the same type"?
3. We generally want small labels (the second parameter), and more detailed messages.

Given those things, I'm thinking something along these lines (feel free to tweak):

```suggestion
                return Err(ShellError::labeled_error(
                    "Not all values can be compared",
                    "not all values compare",
                    tag
                ));
```

The label and message are basically repeated, but that's typical when you don't have any more specific information to include in the message itself. What do you think?

**Bonus round:**

If you're up for it, what if we could include the specific types that don't compare? Knowing the types that aren't comparing gives the user a little more information to work with to resolve the issue.

`coerce_compare` will return an error that is a pair of `str` values naming the types that can't be compared. What if, instead of `all`, you had an iterator that return `Option<(usize, str, str)>`, where the `usize` is the index where things didn't compare, and the two `str` values are the names of the two types. If the result was `Some(...)`, you could produce a much more detailed message then

Happy to give you some guidance on how you could approach this, if desired 🙂

Comment (luccasmmg) [reply crates/nu-cli/src/commands/sort_by.rs:144]: I'm going to change the labes to what you suggested and i'm definitely going to try to give the index and type of the values that can't be coerced

Comment (thegedge) [reply crates/nu-cli/src/commands/sort_by.rs:144]: Sounds good. Reach out to me on discourse if you'd like any tips/help. I'm thinking the end result could be a message along the lines

> Not all values can be compared (<type1> cannot compare against <type2>).

If you'd like, you can make the style/message tweaks in this PR, and you can try the "bonus" in a separate PR. That way we can get your contribution in, which always feels good 🙂

File: crates/nu-cli/src/commands/sort_by.rs [hunk 2]
```diff
@@ -135,6 +135,17 @@ pub fn sort(
...
             let should_sort_case_insensitively = insensitive && vec.iter().all(|x| x.is_string());
 
+            if !vec
+                .windows(2)
+                .all(|elem| coerce_compare(&elem[0], &elem[1]).is_ok())
+            {
+                return Err(ShellError::labeled_error(
+                    "Not all values can be compared",
+                    "not all values compare",
+                    tag
```
Comment (thegedge) [inline crates/nu-cli/src/commands/sort_by.rs:145]: Ugh, my bad, I forgot to mention that in this style our linter expects a comma after all arguments 😞

**Comment by JosephTLyons** [comment]
Maybe we should update the error message here since it technically *shouldn't* be reachable.  Maybe something along the lines of "this will never happen.."
